sematic/ui/packages/common/src/pages/RunDetails/metrics/Metrics.tsx

```diff
@@ -33,42 +34,50 @@ export function TimeseriesMetric(props: {
     const { metricsFilter, color, latestMetricPoints } = props;
     const [payload, , error] = useMetrics(metricsFilter, latestMetricPoints);
 
-    const [chartData, setChartData] = useState<ChartData<"line", number[], string>>({
-        labels: [],
-        datasets: [
-            {
-                label: metricsFilter.metricName,
-                data: [],
-                backgroundColor: color,
-                borderColor: color,
-            },
-        ],
+    const dataStore = useRef({
+        labels: [] as string[],
+        labelSet: new Set<string>(),
+        data: [] as number[],
     });
 
+    const [dataVersion, { inc: bumpDataVersion }] = useCounter();
 
-    useMemo(() => {
+    useEffect(() => {
         if (payload === undefined) return;
-        let labels: string[] = [];
-        let data: number[] = [];
+        const labels: string[] = [];
+        const data: number[] = [];
+        let existingDataReadCursor = 0;
         payload.content.series.forEach((item) => {
             let label = (new Date(item[1][0] * 1000)).toLocaleString();
             labels.push(label.toLocaleString());
-            data.push(item[0]);
+
+            if (dataStore.current.labelSet.has(label)) {
+                data.push(dataStore.current.data[existingDataReadCursor++]);
+            } else {
+                data.push(item[0]);
+                dataStore.current.labelSet.add(label);
+            }
         });
 
-        const cData: ChartData<"line", number[], string> = {
-            labels: labels,
-            datasets: [
-                {
-                    label: metricsFilter.metricName,
-                    data: data,
-                    backgroundColor: color,
-                    borderColor: color,
-                },
-            ],
-        };
-        setChartData(cData);
-    }, [payload, color, metricsFilter.metricName, setChartData]);
+
+        dataStore.current.labels = labels;
+        dataStore.current.data = data;
+
+        bumpDataVersion();
+    }, [payload, bumpDataVersion]);
+
+    const cData: ChartData<"line", number[], string> = useMemo(() => ({
```

Comment:
Now `cData` derives from dataStore, and is updated each time we update dataVersion.